torch/optim/adagrad.py

```diff
@@ -320,11 +320,10 @@ def _multi_tensor_adagrad(
 
     grouped_tensorlists = Optimizer._group_tensors_by_device_and_dtype([params, grads, state_sums, state_steps])
     for ((device_params, device_grads, device_state_sums, device_state_steps), _) in grouped_tensorlists.values():
-
-        device_has_sparse_grad = any(grad.is_sparse for grad in device_grads)
+        device_has_sparse_grad = has_sparse_grad and any(grad.is_sparse for grad in device_grads)
```

Comment:
oh very good heuristical catch

Reply:
Yes, we can also do this for `has_complex`. Being tested in: https://github.com/pytorch/pytorch/pull/110607

Perhaps we should try to cover a smaller surface area with just the Adam changes (https://github.com/pytorch/pytorch/pull/110607) and then proceed with the large change here if things go smoothly?

Reply:
Anw, `has_complex` changes are orthogonal, so I'm also down to just merge this one first and rebase https://github.com/pytorch/pytorch/pull/110607 and further `has_complex` improvements on this PR.